### ISSUE TITLE
(elasticloadbalancingv2): Configure cross-zone load balancing for ALB target groups

### ISSUE THREAD

**Issue by vcschapp** [aws/aws-cdk#31790]
### Describe the feature

Expose ALB's existing ability to turn off cross-zone load balancing for ALB target groups, as described in more detail in the *Other Information* section.

### Use Case

1. As an engineer, I want to build a resilient backend service with control over which availability zones my customer requests are routed to, but I can't because CDK hides ALB's ability to control cross-zone load balancing at the target group level.
2. I'm always frustrated when CDK does not have feature parity with CloudFormation, especially when, as in this case, there is no usable workaround.

### Proposed Solution

The constructor for CDK's [ApplicationTargetGroup](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2.ApplicationTargetGroup.html) construct should accept a `crossZoneEnabled: boolean` flag in the existing constructor props structure, [ApplicationTargetGroupProps](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2.ApplicationTargetGroupProps.html).

#### Feature supported in ALB

Since November 17, 2022, [Application Load Balancers now support turning off cross zone load balancing per target group](https://aws.amazon.com/about-aws/whats-new/2022/11/application-load-balancers-turning-off-cross-zone-load-balancing-per-target-group/). See also the [Cross-zone load balancing](https://docs.aws.amazon.com/elasticloadbalancing/latest/userguide/how-elastic-load-balancing-works.html#cross-zone-load-balancing) heading on the [How Elastic Load Balancing Works](https://docs.aws.amazon.com/elasticloadbalancing/latest/userguide/how-elastic-load-balancing-works.html) user guide page.

#### Feature supported in CloudFormation

This feature can be used in CloudFormation by setting the following target group attribute:

```json
{
  "Key": "load_balancing.cross_zone.enabled",
  "Value": "false"
}
```

under the `TargetGroupAttributes` property within an [AWS::ElasticLoadBalancingV2::TargetGroup](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html#cfn-elasticloadbalancingv2-targetgroup-targetgroupattributes) resource properties.

#### No workaround in CDK

In theory, you can workaround this issue in CDK using the [CfnTargetGroup.addOverride](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2.CfnTargetGroup.html#addwbroverridepath-value). However, because adding overrides would entirely replace the values CDK wants to put for the target group attributes, in practice this means adding overrides would force me to replicate CDK's entire logic for target group attributes. This is incredibly frustrating.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.162.1

### Environment details (OS name and version, etc.)

Ubuntu Linux 20.04

**Comment by khushail** [comment]
Hi @vcschapp , thansk for reaching out.

I see the related issue of configuring cross-zone load balancing false for ALB target groups has been discussed here in detail s-https://github.com/aws/aws-cdk/issues/29866#issuecomment-2062550263

and a corresponding Workaround was also suggested for target groups-https://github.com/aws/aws-cdk/issues/29866#issuecomment-2064577629. 
There is a PR submissions as well- [crossZoneEnabled does not support false for ALB](https://github.com/aws/aws-cdk/pull/29907).

Request you to take a look and get back if this does not solve the usecase you are referring to.

**Comment by vcschapp** [comment]
@khushail thanks.

The issue you referenced is not related (https://github.com/aws/aws-cdk/issues/31790) and the PR you referenced does not fix the problem (https://github.com/aws/aws-cdk/pull/29907).

In issue #31790, the user is requesting to turn off cross-zone at the ALB level, which is not possible.

The [workaround indicated by](https://github.com/aws/aws-cdk/issues/29866#issuecomment-2061508349)  codypenta@ doesn't work for me in general because, as I mentioned in the main issue text:

> In theory, you can workaround this issue in CDK using the [CfnTargetGroup.addOverride](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2.CfnTargetGroup.html#addwbroverridepath-value). However, because adding overrides would entirely replace the values CDK wants to put for the target group attributes, in practice this means adding overrides would force me to replicate CDK's entire logic for target group attributes.

My understanding of the code in [CfnResource.addOverride](https://github.com/aws/aws-cdk/blob/10aa52613e3f5f30c00bb29ad90b28acbd556290/packages/aws-cdk-lib/core/lib/cfn-resource.ts#L236) is that there is no way to supplement keys added by CDK. If you use `addOverride` you are replacing what CDK puts there.

The description for PR https://github.com/aws/aws-cdk/pull/29907 says:

> This PR covers the following cases:
> 
> ...
> 
> 3/ When `crossZoneEnabled` is false with `ALB`, cdk throws an error because ALB does not support disabling it per [doc](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-loadbalancer-loadbalancerattribute.html) description.

So basically the PR makes CDK throw an exception when you try to turn off cross-zone for ALB at the load balancer level. This is fine, as ALB doesn't support that. But that's not what my issue is about. My issue is about disabling cross-zone at the target group level, which ALB does support.

**Comment by khushail** [comment]
@vcschapp , thanks so much for explaining that so well and getting back. 

I understood your point and confirm that its supported by cloudformation as well - https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-targetgroup-targetgroupattribute.html#cfn-elasticloadbalancingv2-targetgroup-targetgroupattribute-key

So till it is implemented , as a workaround, this should be achievable with [Escape hatches](https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html#develop-customize-escape) using L1 construct.

I would be marking this as P2 for keeping it open for contribution from community as well as from the team as well.

Thanks.

**Comment by vcschapp** [comment]
Thanks @khushail, I appreciate the updates.

Btw, it looks like @mazyu36 already has a PR out to potentially fix it!
